store/postgres/lineage_repository.go

```diff
@@ -23,6 +24,278 @@ func NewLineageRepository(client *Client) (*LineageRepository, error) {
 	}, nil
 }
 
-func (repo *LineageRepository) GetEdges(ctx context.Context) ([]lineage.Edge, error) {
-	return nil, nil
+// GetGraph returns a graph that contains list of relations of a given node
+func (repo *LineageRepository) GetGraph(ctx context.Context, node lineage.Node) (lineage.Graph, error) {
+	var graph lineage.Graph
+
+	upstreams, err := repo.getUpstreamsGraph(ctx, node)
+	if err != nil {
+		return graph, fmt.Errorf("error fetching upstreams graph: %w", err)
+	}
+	downstreams, err := repo.getDownstreamsGraph(ctx, node)
+	if err != nil {
+		return graph, fmt.Errorf("error fetching upstreams graph: %w", err)
+	}
+
+	graph = append(upstreams, downstreams...)
+
+	return graph, nil
+}
+
+// Upsert insert or delete connections of a given node by comparing them with current state
+func (repo *LineageRepository) Upsert(ctx context.Context, node lineage.Node, upstreams, downstreams []lineage.Node) error {
+	currentGraph, err := repo.getDirectLineage(ctx, node)
+	if err != nil {
+		return fmt.Errorf("error getting node's direct lineage: %w", err)
+	}
+	newGraph := repo.buildGraph(node, upstreams, downstreams)
+	toInserts, toRemoves := repo.compareGraph(currentGraph, newGraph)
+	toRemoves = repo.filterSelfDeleteOnly(node, toRemoves)
+
+	return repo.client.RunWithinTx(ctx, func(tx *sqlx.Tx) error {
+		err := repo.insertGraph(ctx, tx, toInserts)
+		if err != nil {
+			return fmt.Errorf("error inserting graph: %w", err)
+		}
+
+		err = repo.removeGraph(ctx, tx, toRemoves)
+		if err != nil {
+			return fmt.Errorf("error removing graph: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (repo *LineageRepository) buildGraph(node lineage.Node, upstreams, downstreams []lineage.Node) (graph lineage.Graph) {
```

Comment:
can we move all graph processing functionality to the domain?

Reply:
we could but this one is specific function of the repository implementation (postgres) not a generic one.

if we are making this generic, then it would also makes sense to change the `lineage.Repository` contract so that it receives `lineage.Graph` when upserting as opposed to `[]lineage.Node`.

in summary, it depends on how we are creating our `lineage.Repository` contract, but if we are using the current contract, moving it to domain can give wrong idea.

Reply:
oh yes, that is correct, let's keep it like this then